site/content/docs/rfds/0011-lock-file-target.md

```diff
@@ -0,0 +1,137 @@
+---
+title: Dependency and Lock Files as Targets
+weight: 11
+slug: 0011
+extra:
+  rfd: 11
+  primary_author: Julian Lanson
+  primary_author_link: https://github.com/j-lanson
+  status: Proposed
+---
+
+# Dependency and Lock Files as Targets
+
+Currently, Hipcheck operates on a single-target-per-execution basis. However,
+this model has some friction with one aspect of Hipcheck's elevator pitch,
+namely that you could run it against all the dependencies of your codebase.
+Scripting or manually entering separate Hipcheck invocations for each of your
+project's dependencies puts extra burden on would-be Hipcheck adopters.
+Furthermore, it is computationally wasteful since each invocation would spin
+up, configure, and tear down the same plugin processes that the next invocation
+would also need. To address this, this RFD proposes to make project package and
+lock files that specify dependencies into first-class analysis targets. To do
+so, this document describes proposed changes to the internal API for `Session`
+objects and the `TargetSeed` to `Target` resolution step so that a spun-up set
+of plugins can be re-used for multiple `Target`s in series.
+
+Since projects can have many dependencies, we also propose some parallelization
+strategies. These include potential tweaks to enable multiple Hipcheck instances
+to run simultaneously, or for a single instance to analyze multiple `Targets`
+in parallel with `async` code, despite current restrictions caused by `salsa`.
+
+## TargetSeed to Target Conversion Changes
+
+One of the main API changes introduced by adding package and lock files as
+target types is that we no longer can assume each target supplied to Hipcheck
+resolves to only one repository (a single-repo target). A package file
+containing a list of dependencies would instead correspond to zero or more
+separate repositories (a multi-repo target).
+
+### Command-line Interface Change Considerations
+
+The simple proposed refactor to support this change is to first add new
+`TargetSeedKind` variants to represent package/lock file types. As an example,
+a user may invoke `hc` on the command-line as `hc check Cargo.toml` to
+indicate Hipcheck should run against all dependencies in that file. To make
+logic unique to single- or multi-repo targets easier to write, we probably
+Gshould rename the existing `TargetSeedKind` to `SingleTargetSeed`, add a
+separate `MultiTargetSeed` and then redefine `TargetSeedKind` as follows:
+
+```rust
+enum TargetSeedKind {
+	Single(SingleTargetSeed)
+	Multi(MultiTargetSeed)
+}
+```
+
+There are some types of existing options that do not make sense for a package
+file, such as the `--ref` flag (it is extremely unlikely you would have the same
+target refspec for every repository resolved). Additionally, there are some
+non-existent flags that would make sense for a package target and not for a
+single-repository target; for example, a list of dependencies within the package
+file on which to skip analysis.
```

Comment:
I think there's more to say on refspec issues and it should probably get its own section. Basically yes, the `--ref` flag doesn't make sense for multi-target specifiers, but also some specifiers can embed version / Git ref information. For example, a `Cargo.lock` can have Git info and/or version info, and similarly if we resolve individual package URLs for deps, those can carry version info. So we should make clear that ref-specification is still possible in a multi-target context, just not with the `--ref` flag.